File: src/spec/graph_spec.cpp
#include "simple_pipe/spec/graph_spec.hpp"

#include <fstream>
#include <stdexcept>

namespace simple_pipe {

namespace {

GraphNodeSpec ParseNode(const nlohmann::json& raw) {
  GraphNodeSpec node;
  node.id = raw.at("id").get<std::string>();
  node.type = raw.at("type").get<std::string>();
  if (raw.contains("init_params")) {
    node.init_params = raw.at("init_params");
  } else if (raw.contains("initParams")) {
    node.init_params = raw.at("initParams");
  }
  return node;
}

GraphEdgeSpec ParseEdge(const nlohmann::json& raw) {
  GraphEdgeSpec edge;
  edge.to = raw.at("to").get<std::string>();
  if (raw.contains("from")) {
    const auto& from = raw.at("from");
    if (from.is_string()) {
      edge.from_nodes.push_back(from.get<std::string>());
    } else {
      for (const auto& item : from) {
        edge.from_nodes.push_back(item.get<std::string>());
      }
    }
  } else if (raw.contains("from_nodes")) {
    for (const auto& item : raw.at("from_nodes")) {
      edge.from_nodes.push_back(item.get<std::string>());
    }
  }
  return edge;
}

}  // namespace

std::unordered_set<std::string> PipelineGraphSpec::NodeIds() const {
  std::unordered_set<std::string> ids;
  for (const auto& node : nodes) {
    ids.insert(node.id);
  }
  return ids;
}

const GraphNodeSpec* PipelineGraphSpec::FindNode(const std::string& id) const {
  for (const auto& node : nodes) {
    if (node.id == id) {
      return &node;
    }
  }
  return nullptr;
}

PipelineGraphSpec PipelineGraphSpecFromJson(const nlohmann::json& data) {
  PipelineGraphSpec spec;
  for (const auto& raw : data.at("nodes")) {
    spec.nodes.push_back(ParseNode(raw));
  }
  for (const auto& raw : data.at("edges")) {
    spec.edges.push_back(ParseEdge(raw));
  }
  return spec;
}

PipelineGraphSpec LoadPipelineGraphSpecFromJson(const std::string& path) {
  std::ifstream in(path);
  if (!in) {
    throw std::runtime_error("Failed to open graph spec: " + path);
  }
  nlohmann::json data;
  in >> data;
  return PipelineGraphSpecFromJson(data);
}

}  // namespace simple_pipe

File: simple_pipe/spec/graph_spec.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


@dataclass
class GraphNodeSpec:
    id: str
    type: str
    init_params: dict[str, Any] = field(default_factory=dict)


@dataclass
class GraphEdgeSpec:
    to: str
    from_nodes: list[str] = field(default_factory=list)


@dataclass
class PipelineGraphSpec:
    nodes: list[GraphNodeSpec]
    edges: list[GraphEdgeSpec]

    def node_ids(self) -> set[str]:
        return {n.id for n in self.nodes}

    def node_by_id(self, node_id: str) -> GraphNodeSpec | None:
        for node in self.nodes:
            if node.id == node_id:
                return node
        return None


def _parse_node(raw: dict[str, Any]) -> GraphNodeSpec:
    return GraphNodeSpec(
        id=raw["id"],
        type=raw["type"],
        init_params=dict(raw.get("init_params") or raw.get("initParams") or {}),
    )


def _parse_edge(raw: dict[str, Any]) -> GraphEdgeSpec:
    upstream = raw.get("from") or raw.get("from_nodes") or []
    if isinstance(upstream, str):
        upstream = [upstream]
    return GraphEdgeSpec(to=raw["to"], from_nodes=list(upstream))


def pipeline_graph_spec_from_dict(data: dict[str, Any]) -> PipelineGraphSpec:
    nodes = [_parse_node(n) for n in data.get("nodes", [])]
    edges = [_parse_edge(e) for e in data.get("edges", [])]
    return PipelineGraphSpec(nodes=nodes, edges=edges)


def load_graph_spec_from_json(path: str | Path) -> PipelineGraphSpec:
    text = Path(path).read_text(encoding="utf-8")
    return pipeline_graph_spec_from_dict(json.loads(text))
